### DISCUSSION TITLE
Creation of diagrams during static side build using mermaidjs (or equivalent)

### DISCUSSION THREAD

**Discussion by sandys** [vercel/next.js#12837] [Help]
hi guys,
I'm looking to build documentation sites that has diagrams. Similar to the ones generated by Mermaidjs - https://mermaid-js.github.io/mermaid/#/

But i want to generate these diagrams at build time from markup. Does Nextjs have any such solution ?

**Comment by Timer** [comment]
You can write any Node.js code you'd like in `getStaticProps`, including generating diagrams.

**Comment by Izhaki** [comment]
You can add mermaid by adding it to `_document.js` in the `Head` block:

```
        <Head>
          <script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
          <script
            // eslint-disable-next-line react/no-danger
            dangerouslySetInnerHTML={{
              __html: `mermaid.initialize({startOnLoad: true});`,
            }}
          />
        </Head>
```

Then either in React or an MDX document just put:

```
  <div class="mermaid">
graph LR
A-->B
  </div>
```

That's how mermaid works.

-----

You can also use it's low-level API, the component will look something like this:

```jsx
import React from 'react';

let currentId = 0;
const uuid = () => `mermaid-${(currentId++).toString()}`;

function Mermaid({ graphDefinition }) {
  const [html, setHtml] = React.useState('');
  React.useLayoutEffect(() => {
    if (graphDefinition) {
      try {
        window.mermaid.mermaidAPI.render(uuid(), graphDefinition, svgCode =>
          setHtml(svgCode)
        );
      } catch (e) {
        setHtml('')
        console.error(e);
      }
    }
  }, [graphDefinition]);

  return graphDefinition ? <div dangerouslySetInnerHTML={{ __html: html }} /> : null;
}

export default Mermaid;
```

Note that this uses `window.mermaid`, so you still need this in `_document.js`:

```
          <script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
```

But you can skip the `mermaid.initialize` part.

**Reply by TomasHubelbauer** [reply]
@lzhaki any change you'd know how to adopt this for Nextra? I tried copy-pasting as-is assuming stuff like `Head` would work in there since it is based on Next but that doesn't seem to be the case. My Mermaid code still shows up as Mermaid.

**Reply by TomasHubelbauer** [reply]
Nevermind, this should be coming to Nextra soon according to https://github.com/shuding/nextra/issues/1373
